packages/toolpad-app/src/toolpad/Apps/index.tsx

```diff
@@ -199,23 +199,55 @@ function CreateAppDialog({ onClose, open, ...props }: CreateAppDialogProps) {
 
         sendAppCreatedEvent(createdApp.name, appTemplateId);
       } catch (rawError) {
-        if (config.recaptchaV2SiteKey && !hasShownRecaptchaCheckbox && captchaTargetRef.current) {
-          const error = errorFrom(rawError);
-          if (error.code === ERR_VALIDATE_CAPTCHA_FAILED) {
-            // Show captcha checkbox
-            const widgetId = grecaptcha.render(captchaTargetRef.current.id, {
-              sitekey: config.recaptchaV2SiteKey,
-            });
-            latestRecaptchaWidgetIdRef.current = widgetId;
-          }
+        const error = errorFrom(rawError);
+        if (config.recaptchaV2SiteKey && error.code === ERR_VALIDATE_CAPTCHA_FAILED) {
+          // Show captcha checkbox
+          setRequestRecaptchaV2(true);
+        } else {
+          throw error;
         }
       }
     },
-    [appTemplateId, createAppMutation, dom, isFormValid, name, setLatestCreatedApp],
+    [
+      appTemplateId,
+      createAppMutation,
+      dom,
+      isFormValid,
+      name,
+      recaptchaV2Token,
+      requestRecaptchaV2,
+      setLatestCreatedApp,
+    ],
   );
 
   return (
     <Dialog {...props} open={open} onClose={config.isDemo ? NO_OP : onClose} maxWidth="xs">
+      {config.recaptchaV3SiteKey ? (
+        <Script
+          strategy="afterInteractive"
+          src={`https://www.google.com/recaptcha/api.js?render=${encodeURIComponent(
+            config.recaptchaV3SiteKey,
+          )}`}
+          onLoad={async () => {
```

Comment:
So we're always rendering but we still only show it if the V3 captcha validation fails, seems good.
We could move have this function outside the JSX if you want, just to separate the HTML from the logic a bit more.

Reply:
I noticed that if we close the "Create app" modal and open it again the space for the visual captcha isn't there anymore, and the captcha doesn't show? Maybe the `id` and `getResponse` were handling something about this? - I can take a deeper look if you want.

Reply:
Should probably use `display` instead of `visibility`